### PR TITLE
Let the choice to write credential for HTTP source on external dictionaries

### PR DESCRIPTION
**Pull request by YiuRULE**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

Set two configuration options for a dictionary based on an HTTP source. `credentials` and `http-headers`.

Detailed description (optional):

Two options has been made, mostly for improve security when we use a private dataset using an HTTP dictionary.

The first one is `credentials`, used when a server use a basic http auth. With an user and a password.

The second is `http-headers` for set some datas on the HTTP request header. The motivation of it is for service who can possibly get an API key from an header, to be able let ClickHouse get the datas.

### CODE REVIEW

File: dbms/src/Dictionaries/HTTPDictionarySource.cpp [hunk 1]
```diff
@@ -34,19 +34,45 @@ HTTPDictionarySource::HTTPDictionarySource(
     , context(context_)
     , timeouts(ConnectionTimeouts::getHTTPTimeouts(context))
 {
+    const auto & credentials_prefix = config_prefix + ".credentials";
+
+    if (config.has(credentials_prefix))
+    {
+        this->credentials.setUsername(config.getString(credentials_prefix + ".user", ""));
```
Comment (alexey-milovidov) [inline dbms/src/Dictionaries/HTTPDictionarySource.cpp]: this-> qualification looks unneeded.

File: dbms/src/IO/ReadWriteBufferFromHTTP.h [hunk 2]
```diff
@@ -86,6 +86,10 @@ namespace detail
     template <typename UpdatableSessionPtr>
     class ReadWriteBufferFromHTTPBase : public ReadBuffer
     {
+    public:
+        using HttpHeaderEntry = std::tuple<std::string, std::string>;
```
Comment (alexey-milovidov) [inline dbms/src/IO/ReadWriteBufferFromHTTP.h]: We prefer HTTP, not Http.

File: docs/en/query_language/dicts/external_dicts_dict_sources.md [hunk 3]
```diff
@@ -84,6 +84,16 @@ Example of settings:
     <http>
         <url>http://[::1]/os.tsv</url>
         <format>TabSeparated</format>
+        <credentials>
```
Comment (alexey-milovidov) [inline docs/en/query_language/dicts/external_dicts_dict_sources.md]: Need to mention that it's optional.

File: docs/en/query_language/dicts/external_dicts_dict_sources.md [hunk 4]
```diff
@@ -84,6 +84,16 @@ Example of settings:
...
+            <user>user</user>
+            <password>password</password>
+        </credentials>
+        <http-headers>
```
Comment (alexey-milovidov) [inline docs/en/query_language/dicts/external_dicts_dict_sources.md]: Maybe we can omit `http-` here because it's http source.

**Review by alexey-milovidov — APPROVED** [review]
Ok. Only a few style issues remain.

File: docs/en/query_language/dicts/external_dicts_dict_sources.md [hunk 5]
```diff
@@ -84,6 +84,16 @@ Example of settings:
     <http>
         <url>http://[::1]/os.tsv</url>
```
Comment (alexey-milovidov) [inline docs/en/query_language/dicts/external_dicts_dict_sources.md]: Why we cannot include credentials in URL:
http://user:password@[::1]/os.tsv

Comment (YiuRULE) [reply docs/en/query_language/dicts/external_dicts_dict_sources.md]: I suspect a bit the code who handle authentication to not handle it if we set the auth on the `URI`, I tried to do some test locally with an HTTP server, the server returned a `401` error if we put the auth on the URL.

https://github.com/ClickHouse/ClickHouse/blob/master/dbms/src/IO/ReadWriteBufferFromHTTP.h#L112-L113

**Comment by YiuRULE** [comment]
Ok done :)

**Comment by alexey-milovidov** [comment]
Ok. Now it's almost ready!
We also need to update integration test to ensure that this feature continue to work.

Look at `tests/integration//test_redirect_url_storage`

or 

`tests/queries/0_stateless/00646_url_engine.python`

**Comment by YiuRULE** [comment]
Done ! :)

I modified one integration test where the http server do a simili-authentication system

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**